src/main/java/frc/robot/systems/IntakeFSMSystem.java

```diff
@@ -0,0 +1,707 @@
+					return IntakeFSMState.MOVE_TO_HOME;
+				} else {
+					return IntakeFSMState.MOVE_TO_GROUND;
+				}
+
+			case OUTTAKING:
+				if (inRange(throughBore.getDistance(), Constants.GROUND_ENCODER_COUNT)
+					&& input.isOuttakeButtonPressed() && !input.isIntakeButtonPressed()
+					&& !input.isShootButtonPressed()) {
+					return IntakeFSMState.OUTTAKING;
+				}
+
+				if (input.isShootButtonPressed()
+					&& !(input.isOuttakeButtonPressed() || input.isIntakeButtonPressed())) {
+					return IntakeFSMState.MOVE_TO_HOME;
+				} else {
+					return IntakeFSMState.MOVE_TO_GROUND;
+				}
+
+			default:
+				throw new IllegalStateException("Invalid state: " + currentState.toString());
+		}
+	}
+
+/* ------------------------ Command handlers ------------------------ */
+
+	private boolean inRange(double targetEncoder, double currentEncoder) {
+		return Math.abs(targetEncoder - currentEncoder) < Constants.INRANGE_VALUE;
+	}
+
+	/**
+	 * A PID controller for the pivot motor.
+	 * @param currentEncoderPID The current position of the pivot motor in encoder units.
+	 * @param targetEncoder The target position of the pivot motor in encoder units.
+	 * @return The correct power to send to the pivot motor to achieve the target position.
+	 */
+	private double pid(double currentEncoderPID, double targetEncoder) {
+		double correction = Constants.PID_CONSTANT_PIVOT_P
+			* (targetEncoder - currentEncoderPID);
+		return Math.min(Constants.MAX_TURN_SPEED,
+			Math.max(Constants.MIN_TURN_SPEED, correction));
+	}
+
+	/**
+	 * A PID controller for the pivot motor in autonomous mode.
+	 * @param currentEncoderPID The current position of the pivot motor in encoder units.
+	 * @param targetEncoder The target position of the pivot motor in encoder units.
+	 * @return The correct power to send to the pivot motor to achieve the target position.
+	 */
+	private double pidAuto(double currentEncoderPID, double targetEncoder) {
+		double correction = Constants.PID_CONSTANT_PIVOT_P_AUTO
+			* (targetEncoder - currentEncoderPID);
+		return Math.min(Constants.MAX_TURN_SPEED_AUTO,
+			Math.max(Constants.MIN_TURN_SPEED_AUTO, correction));
+	}
+
+	/**
+	 * Checks if the intake is holding a note.
+	 * @return if the intake is holding a note
+	 */
+	public boolean hasNote() {
+		boolean isInRange = colorSensor.getProximity() >= Constants.PROXIMIIY_THRESHOLD;
+		SmartDashboard.putBoolean("is close enough", isInRange);
+
+		if (isInRange) {
+			noteColorFrames++;
+		} else {
+			noteColorFrames = 0;
+		}
+		hasNote = noteColorFrames >= Constants.NOTE_FRAMES_MIN;
+
+		return hasNote;
+	}
+
+	/* ------------------------ FSM state handlers ------------------------ */
+	/**
+	 * Handle behavior in MOVE_TO_HOME.
+	 * @param input Global TeleopInput if robot in teleop mode or null if
+	 *        the robot is in autonomous mode.
+	 */
+	private void handleMoveHomeState(TeleopInput input) {
+		if (hasNote) {
+			led.blueLight();
```

Comment:
Is there a general description somewhere of what all the colors mean? Some quick reference we can hand to the drivers.

Reply:
@beranki might be able to answer that one, I am not aware if we ever drafted these out on a document